autoendpoint/src/routers/fcm/router.rs

```diff
@@ -163,36 +165,43 @@ impl Router for FcmRouter {
         // that is sent by the client at registration.
         //
         // Try reading as FCM and fall back to GCM.
+
         let (routing_token, app_id) =
             self.routing_info(router_data, &notification.subscription.user.uaid)?;
         let ttl = MAX_TTL.min(self.settings.min_ttl.max(notification.headers.ttl as usize));
-        let message_data = build_message_data(notification)?;
 
         // Send the notification to FCM
-        // (Sigh, errors do not have tags support. )
         let client = self
             .clients
             .get(&app_id)
             .ok_or_else(|| FcmError::InvalidAppId(app_id.clone()))?;
 
-        // GCM is the older message format for android, and it's not possible to generate
-        // new test keys. As of 2023-09, GCM should use FCM OAuth2 in order to authenticate.
-        // FCM operates by using the more complex token as part of an OAuth2
-        // transaction. According to the documentation, GCM and FCM are interoperable,
-        // meaning that the client provided tokens should match up, and that the structure
-        // of the data should also not make a huge difference.
-        //
-        // In Oct of 2023, "legacy" GCM support was deprecated. After some
-        // research, it was hypothesised that the current FCM project may be able
-        // to accept GCM routing identifiers as FCM message tokens. The current code presumes
-        // that this is a fragile relationship and so GCM errors do not immediately drop the
-        // remote user's endpoint (potentially causing a UA reset.)
-        //
-        let platform = if client.is_gcm {
-            "gcm_as_fcmv1"
-        } else {
-            "fcmv1"
-        };
+        let router_type = RouterType::from_str(&notification.subscription.user.router_type)
+            .unwrap_or(RouterType::FCM);
+        if matches!(router_type, RouterType::GCM) {
+            // GCM is the older message format for android, and it's not possible to generate
+            // new test keys.
+            // As of 2023-09-22 Legacy GCM messages are no longer supported.
+            // Due to a very unfortunate set of decisions made at the beginning of the WebPush
+            // project, we do not have access to the original GCM credential set and therefore
+            // can not move these GCM users to FCM. We have no choice but to treat this as
+            // a bridge server rejection, disable the subscription, and hope that
+            // the remote UA is running Fenix+, which will perform a daily check-in
```

Comment:
Would Fenix+ have any GCM subscriptions? Per our earlier discussion it sounded like Fenix+ would have been fully GCM?

Reply:
It shouldn't have any GCM subscriptions, no. As you are well aware, however, I tend to put my faith in Murphy, and if there's even the most remote chance that some record might have this crap in it somewhere, or something slips through the cracks in some weird clippy way, I'm fine leaving a second line of defense in place for now and pulling it later when we're sure that it hasn't been called. 

Anything that can be killed can be overkilled, and sometimes, that's exactly the right amount.

Reply:
I'd argue the comment's a bit misleading then

Reply:
We don't need this code block either with `validate_user` doing it up front -- why keep both?